text/0000-const-type-inference.md

```diff
@@ -0,0 +1,124 @@
+- Code generators may not have enough information to easily determine the type.
+
+This change aims to make Rust code more concise and maintainable, especially in scenarios where the types of
+const items are complicated or not easily expressible.
+
+Inferring constant types also improves the ergonomics of the language, particularly for new users. Some users are
+coming from languages where most (or all) types are inferred, so inferring obvious types matches their
+expectations. Other new users are focused on learning ownership, or other core Rust concepts. Reducing the
+amount of boilerplate reduces their mental load. This reduction in mental load also helps experienced programmers.
+
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+You may declare constants and static variables without specifying their types when the type can be inferred
+from the initial value. For example:
+
+```rs
+const PI = 3.1415; // inferred as f64
+static MESSAGE = "Hello, World!"; // inferred as &'static str
+const FN_PTR = std::string::String::default; // inferred as fn() -> String
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+
+The type inference for `const` and `static` will leverage Rust's existing type inference mechanisms. The compiler will infer the type exclusively based on the RHS. If the type cannot be determined or if it leads to ambiguities, the compiler will emit an error, prompting the programmer to specify the type explicitly.
+
+
+Today, the compiler already gives hint for most cases where the const or static item is missing a type:
+
+```
+802 | const A = 0;
+    |        ^ help: provide a type for the constant: `: i32`
+```
+
+
+```
+error: missing type for `const` item                                                     
+  --> file.rs:27:26
+   |
+27 |     pub const update_blas = SystemStage { system: test_system, stage: vk::Pipeli... 
+   |                          ^ help: provide a type for the constant: `: render_pass::SystemStage<for<'a> fn(ResMut<'a, AsyncQueues>)>`
+```
+
+The implementation should only need to carry over this information and set the type correspondingly
+instead of emitting an error.
+
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+- Potential Loss of Clarity: In some cases, omitting the type might make the code less clear,
+  especially to newcomers or when explicit types are needed to understanding the purpose of the item.
+  It is my belief that this is a choice better left for the developers as in the case of `let` bindings.
+- Semvar compatibilty: It's a good idea that public API endpoints should be "obviously semvar stable".
+  However, not all `const` or `static` items are public, and explicit typing isn't always important for semvar stability.
+  Requiring explicit typing for this reason seems a bit heavy handed.
+
+Both of these drawback could be addressed using an allow-by-default clippy lint for `const` and `static` types.
```

Comment:
I don't think an allow-by-default lint addresses concerns about it being potentially confusing. Having a lint can be useful, but it doesn't address the problem because almost everyone won't be using it (but may be using inference).

Reply:
Fair enough, happy to withdraw this suggestion and go with what you've suggested instead.